docs(dh): add docs for feature flags (#679)
Energinet-DataHub/greenforce-frontend: libs/dh/shared/feature-flags/README.md

# DataHub feature flags

You can disable functionality for specified environments: `dev, test, prod.`

<mark>Notice: feature flags treat the pre-prod environment as prod.</mark>

Feature flags are not supposed to live long; therefore, they have a **maximum lifetime of 14 days** (If a feature flag is older than 14 days, an automated test will fail).

## Create a feature flag

Feature flags are located here: `libs/dh/shared/feature-flags/src/lib/feature-flags.ts` -> Find the constant: `dhFeatureFlagsConfig`

The feature flags require two properties, `created` and `disabledEnvironments`.

**Example:**

````ts
'my-awesome-feature-flag': {
  created: '01-01-2022',
  disabledEnvironments: [DhAppEnvironment.prod],
}

The `created` property is danish locale `dd-mm-yyyy`. Feature flags use this property internally to check for old feature flags.

The `disabledEnvironments` property is an array of the environments you want to disable a feature in, e.g. `prod`.

## Using a feature flag

Feature flags can be used by the `*dhFeatureflag` directive or the `DhFeatureFlagsService.`

> <mark>Notice: Unknown feature flags not added to the feature flag configuration are treated as enabled.</mark>

Author: @ManBearTM You're absolutely right unless you "cheat" with `as DhFeatureFlags`. This is ofc. not the intended usage but the behavior would be as mentioned in the doc. However, we could skip documenting this, as it's not intended usage what do you think?

> <mark>Notice: Unknown feature flags not added to the feature flag configuration are treated as enabled.</mark>

Author: Actually there's also the use-case where no feature flags are added to the configuration. then `DhFeatureFlags` will get the type `never`.

### From the template

Use the feature flags within templates with the feature flag structural directive.

```html
<ng-container *dhFeatureFlag="my-awesome-feature-flag">
  SOME CONTENT
</ng-container>
````

<mark>Notice: Using feature flags with the structural directive, you will need to import the `DhFeatureFlagDirectiveModule`</mark>

### From TypeScript (Controllers, Guards, etc.)

Use the feature flags within TypeScript with the feature flag service.

1. Inject the service (The service is provided in the root, so you don't have to import any module):

```ts
  constructor(private featureFlagsService: DhFeatureFlagsService) {}
```

2. Use the `isEnabled` method on the service:

```ts
this.featureFlagsService.isEnabled('my-awesome-feature-flag');

> ```

Reviewer: Perhaps add a short section about how to remove the flags again, something like this:
```md
## Remove a feature flag

When it is time to remove a feature flag, it is recommended to begin by removing the entry from
`dhFeatureFlagsConfig`. The compiler will then error in all the files using that feature flag, making
it easier to identify which files need to be cleaned up.
```

Author: Good point!

# DataHub feature flags

You can disable functionality for specified environments: `dev, test, prod.`

<mark>Notice: feature flags treat the pre-prod environment as prod.</mark>

Feature flags are not supposed to live long; therefore, they have a **maximum lifetime of 14 days** (If a feature flag is older than 14 days, an automated test will fail).

## Create a feature flag

Feature flags are located here: `libs/dh/shared/feature-flags/src/lib/feature-flags.ts` -> Find the constant: `dhFeatureFlagsConfig`

The feature flags require two properties, `created` and `disabledEnvironments`.

**Example:**

> ```ts
> 'example-feature-flag': {
>     created: '01-01-2022',
>     disabledEnvironments: [DhAppEnvironment.prod],
>   }
> ```

Reviewer 2: nitpick:suggestion: let's use the same feature flag name mentioned later in the doc.

```suggestion
```ts
'my-awesome-feature-flag': {
  created: '01-01-2022',
  disabledEnvironments: [DhAppEnvironment.prod],
}
```
```

The `created` property is danish locale `dd-mm-yyyy`. Feature flags use this property internally to check for old feature flags.

> The `disabledEnvironments` property is an array of the environments you wan't to disable the feature, for ex., `prod.`

Reviewer 2: suggestion:

```suggestion
The `disabledEnvironments` property is an array of the environments you want to disable a feature in, e.g. `prod`.
```
